File: mahaam-api-py/feat/user/device_repo.py
from typing import Protocol
from uuid import UUID, uuid4
from typing import List
from infra.db import DB
from feat.user.user_model import Device
from infra.validation import ProtocolEnforcer

class DeviceRepo(Protocol):
    def create(self, device: Device, conn) -> UUID: ...
    def delete(self, id: UUID, conn=None) -> int: ...
    def delete_by_user(self, user_id: UUID, except_device_id: UUID) -> int: ...
    def delete_by_fingerprint(self, fingerprint: str, conn) -> int: ...
    def select_one(self, id: UUID) -> Device | None: ...
    def select_many(self, user_id: UUID, conn=None) -> List[Device]: ...
    def update_user_id(self, device_id: UUID, user_id: UUID, conn) -> int: ...

class DefaultDeviceRepo(metaclass=ProtocolEnforcer, protocol=DeviceRepo):
    def create(self, device: Device, conn) -> UUID:
        device_id = uuid4()
        sql = """
        INSERT INTO devices (id, user_id, platform, fingerprint, info, created_at) 
        VALUES (:id, :user_id, :platform, :fingerprint, :info, current_timestamp)"""
        params = {
            "id": str(device_id),
            "user_id": str(device.user_id),
            "platform": device.platform,
            "fingerprint": device.fingerprint,
            "info": device.info
        }
        DB.insert(sql, params, conn)
        return device_id

    def delete(self, id: UUID, conn=None) -> int:
        sql = "DELETE FROM devices WHERE id = :id"
        return DB.delete(sql, {"id": str(id)}, conn)

    def delete_by_user(self, user_id: UUID, except_device_id: UUID) -> int:
        sql = "DELETE FROM devices WHERE user_id = :user_id AND id != :except_device_id"
        return DB.delete(sql, {"user_id": str(user_id), "except_device_id": str(except_device_id)})

    def delete_by_fingerprint(self, fingerprint: str, conn) -> int:
        sql = "DELETE FROM devices WHERE fingerprint = :fingerprint"
        return DB.delete(sql, {"fingerprint": fingerprint}, conn)

    def select_one(self, id: UUID) -> Device | None:
        sql = """
            SELECT id, user_id, platform, fingerprint, info, created_at
            FROM devices WHERE id = :id ORDER BY created_at DESC"""
        return DB.select_one(Device, sql, {"id": str(id)})

    def select_many(self, user_id: UUID, conn=None) -> List[Device]:
        sql = """
            SELECT id, user_id, platform, fingerprint, info, created_at
            FROM devices WHERE user_id = :user_id ORDER BY created_at DESC"""
        return DB.select_many(Device, sql, {"user_id": str(user_id)}, conn)

    def update_user_id(self, device_id: UUID, user_id: UUID, conn) -> int:
        sql = "UPDATE devices SET user_id = :user_id, updated_at = current_timestamp WHERE id = :device_id"
        return DB.update(sql, {"device_id": str(device_id), "user_id": str(user_id)}, conn)

File: mahaam-api-java/src/main/java/mahaam/feat/user/DeviceRepo.java
package mahaam.feat.user;

import java.util.List;
import java.util.UUID;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import mahaam.feat.user.UserModel.Device;
import mahaam.infra.DB;
import mahaam.infra.Log;
import mahaam.infra.Mapper;

public interface DeviceRepo {
	UUID create(Device device);

	int delete(UUID id);

	int deleteByFingerprint(String fingerprint);

	Device getOne(UUID id);

	List<Device> getMany(UUID userId);

	int updateUserId(UUID deviceId, UUID userId);
}

@ApplicationScoped
class DefaultDeviceRepo implements DeviceRepo {
	@Inject
	DB db;

	@Override
	public UUID create(Device device) {
		try {
			String query = """
					INSERT INTO devices (id, user_id, platform, fingerprint, info, created_at)
					VALUES (:id, :userId, :platform, :fingerprint, :info, current_timestamp)""";
			var deviceId = UUID.randomUUID();
			var params = Mapper.of(
					"id",
					deviceId,
					"userId",
					device.userId,
					"platform",
					device.platform,
					"fingerprint",
					device.fingerprint,
					"info",
					device.info);
			db.insert(query, params);
			return deviceId;
		} catch (Exception e) {
			Log.error("Failed to create device: " + e.getMessage());
			throw new RuntimeException("Failed to create device", e);
		}
	}

	@Override
	public int delete(UUID id) {
		String query = "DELETE FROM devices WHERE id = :id";
		return db.delete(query, Mapper.of("id", id));
	}

	@Override
	public int deleteByFingerprint(String fingerprint) {
		String query = "DELETE FROM devices WHERE fingerprint = :fingerprint";
		return db.delete(query, Mapper.of("fingerprint", fingerprint));
	}

	@Override
	public Device getOne(UUID id) {
		String query = """
				SELECT id d_id, user_id d_userId, platform d_platform, fingerprint d_fingerprint, info d_info, created_at d_createdAt
				FROM devices d WHERE d.id = :id order by d.created_at desc""";
		return db.selectOne(query, Device.class, Mapper.of("id", id));
	}

	@Override
	public List<Device> getMany(UUID userId) {
		String query = """
				SELECT id d_id, user_id d_userId, platform d_platform, fingerprint d_fingerprint, info d_info, created_at d_createdAt
				FROM devices d WHERE d.user_id = :userId order by d.created_at desc""";
		return db.selectList(query, Device.class, Mapper.of("userId", userId));
	}

	@Override
	public int updateUserId(UUID deviceId, UUID userId) {
		String query = "UPDATE devices SET user_id = :userId, updated_at = current_timestamp WHERE id = :deviceId";
		return db.update(query, Mapper.of("deviceId", deviceId, "userId", userId));
	}
}
